pkg/converter/tool/feature_test.go

```diff
@@ -7,41 +7,757 @@
+			require.Equal(t, tt.expect, tt.features)
+		})
+	}
+
+	testsContains := []struct {
+		name     string
+		features Features
+		item     Feature
+		expect   bool
+	}{
+		{
+			name:     "should return contains",
+			features: Features{FeatureBatchSize: {}, FeatureTar2Rafs: {}},
+			item:     FeatureTar2Rafs,
+			expect:   true,
+		},
+		{
+			name:     "should return not contains",
+			features: Features{FeatureBatchSize: {}},
+			item:     FeatureTar2Rafs,
+			expect:   false,
+		},
+		{
+			name:     "Contains should accept empty string",
+			features: Features{FeatureBatchSize: {}},
+			item:     "",
+			expect:   false,
+		},
+	}
+	for _, tt := range testsContains {
+		t.Run(tt.name, func(t *testing.T) {
+			require.Equal(t, tt.expect, tt.features.Contains(tt.item))
+		})
+	}
+
+	testsEquals := []struct {
+		name     string
+		features Features
+		other    Features
+		expect   bool
+	}{
+		{
+			name:     "should successfully check equality",
+			features: Features{FeatureBatchSize: {}, FeatureTar2Rafs: {}},
+			other:    Features{FeatureBatchSize: {}, FeatureTar2Rafs: {}},
+			expect:   true,
+		},
+		{
+			name:     "should successfully check inequality with different length",
+			features: Features{FeatureBatchSize: {}, FeatureTar2Rafs: {}},
+			other:    Features{FeatureBatchSize: {}},
+			expect:   false,
+		},
+		{
+			name:     "should successfully check inequality with different items",
+			features: Features{FeatureTar2Rafs: {}},
+			other:    Features{FeatureBatchSize: {}},
+			expect:   false,
+		},
+		{
+			name:     "should ignore order",
+			features: Features{FeatureBatchSize: {}, FeatureTar2Rafs: {}},
+			other:    Features{FeatureTar2Rafs: {}, FeatureBatchSize: {}},
+			expect:   true,
+		},
+		{
+			name:     "Equals should accept nil",
+			features: Features{FeatureBatchSize: {}},
+			other:    nil,
+			expect:   false,
+		},
+	}
+	for _, tt := range testsEquals {
+		t.Run(tt.name, func(t *testing.T) {
+			require.Equal(t, tt.expect, tt.features.Equals(tt.other))
+		})
+	}
+
+	testsIsLegal := []struct {
+		name    string
+		feature Feature
+		expect  bool
+	}{
+		{
+			name:    "hard-coded feature '--batch-size' should be legal",
+			feature: FeatureBatchSize,
+			expect:  true,
+		},
+		{
+			name:    "hard-coded feature '--type tar-rafs' should be legal",
+			feature: FeatureTar2Rafs,
+			expect:  true,
+		},
+		{
+			name:    "IsLegal should accept empty string",
+			feature: "",
+			expect:  false,
+		},
+		{
+			name:    "'[' should be illegal since it is a symbol of regular expression",
+			feature: "[",
+			expect:  false,
+		},
+	}
+	for _, tt := range testsIsLegal {
+		t.Run(tt.name, func(t *testing.T) {
+			require.Equal(t, tt.expect, tt.feature.IsLegal())
+		})
+	}
+}
+
+func TestDetectFeature(t *testing.T) {
+	tests := []struct {
+		name    string
+		feature Feature
+		helpMsg []byte
+		expect  bool
+	}{
+		{
+			name:    "'--type tar-rafs' is supported in v2.2.0-239-gf5c08fcf",
```

Comment:
Very detailed test! But I'm worried about how we keep compatible if the `nydus-image` output changes.

Reply:
@imeoer The possible way to change `nydus-image create -h` output I've found:

1. Change the args of features, e.g., `--source-type directory` in v2.1.4 to `--type directory` in v2.2.0. It can be recognized by adding a `featureMap` to check multiple aliases of one feature, if needed in the future:
```golang
var featureMap = map[Feature][]string{
	FeatureTar2Rafs: {"--type tar-rafs", "--source-type tar"},
}
```
2. Deprecate the args of features, e.g.,
```shell
	--backend-type <backend-type>
		[deprecated!] Blob storage backend type, only support localfs for compatibility. Try use --blob instead.
		[possible values: localfs]
```
Deprecating an arg adds an additional line in help message starting with `[deprecated!]`, and have been handled and tested in `detectFeature()`.

3. Hide the args of features. I guess it can be used for experimental features, deprecated features, or some features that don't need to be considered by end users. For experimental features, we cannot hide them anymore in the future. For features we need to deprecate, we should only deprecate them rather than hide them to provide end users clear boundaries about when the args are deprecated and removed. For internal features, if the feature detection is needed, it's true that just check the help message is not enough.
4. Update the CLI dependences of `nydus-image`. It should be compatible.

Could you please provide me with some specific cases on how the `nydus-image` output would possibly change?

Reply:
I think we can just simply check the keywords `--type` and `tar-rafs` for feature `--type tar-rafs`. :)

Reply:
Changed.